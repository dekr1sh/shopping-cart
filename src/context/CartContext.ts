import { createContext } from "react"
import type { Product } from "../data/products.ts"  // specifying the 'type' keyword optimizes the build

export interface CartItem extends Product {
  size: string       // shoe size
  quantity: number   // number of shoes
}

interface CartContextType {
  cartItems: CartItem[]
  addToCart: (product: Product, size: string, quantity: number) => void
  removeFromCart: (id: number, size: string) => void
  increaseQuantity: (id: number, size: string) => void
  decreaseQuantity: (id: number, size: string) => void
  clearCart: () => void
}

export const CartContext = createContext<CartContextType | undefined>(undefined)
