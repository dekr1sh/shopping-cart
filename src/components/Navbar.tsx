import { Link, useLocation } from "react-router-dom";
import { ShoppingBag, Menu, X } from "lucide-react";
import { Button } from "@/components/ui/button";
import { useCart } from "@/context/useCart";
import { useState, useEffect, useRef } from "react";
import CartDropdown from "@/components/CartDropdown";

export default function Navbar() {
  const { cartItems } = useCart();
  const [isCartOpen, setIsCartOpen] = useState(false);
  const [isMobileMenuOpen, setIsMobileMenuOpen] = useState(false);
  const cartRef = useRef<HTMLDivElement>(null);
  const location = useLocation();

  const totalItems = cartItems.reduce((total, item) => total + item.quantity, 0);

  useEffect(() => {
    const handleClickOutside = (event: MouseEvent) => {
      if (cartRef.current && !cartRef.current.contains(event.target as Node)) {
        setIsCartOpen(false);
      }
    };

    document.addEventListener("mousedown", handleClickOutside);
    return () => {
      document.removeEventListener("mousedown", handleClickOutside);
    };
  }, []);

  // Close mobile menu when route changes
  useEffect(() => {
    setIsMobileMenuOpen(false);
  }, [location.pathname]);

  return (
    <header className="sticky top-0 z-50 bg-background/80 backdrop-blur-md border-b container mx-auto px-4">
      <div className="flex items-center justify-between h-16">
        {/* Logo */}
        <Link to="/" className="font-bold text-xl flex items-center">
          <span className="text-primary">Drip</span>Kicks
        </Link>

        {/* Mobile Menu Button */}
        <button
          className="md:hidden"
          onClick={() => setIsMobileMenuOpen(!isMobileMenuOpen)}
          aria-label={isMobileMenuOpen ? "Close menu" : "Open menu"}
        >
          {isMobileMenuOpen ? <X size={24} /> : <Menu size={24} />}
        </button>

        {/* Desktop Navigation */}
        <nav className="hidden md:flex items-center space-x-8">
          <Link
            to="/"
            className={`hover:text-primary transition-colors ${location.pathname === "/" ? "text-primary font-medium" : ""}`}
          >
            Home
          </Link>
          <Link
            to="/shop"
            className={`hover:text-primary transition-colors ${location.pathname === "/shop" ? "text-primary font-medium" : ""}`}
          >
            Shop
          </Link>
        </nav>

        {/* Cart Button */}
        <div className="relative" ref={cartRef}>
          <Button
            variant="ghost"
            size="icon"
            className="relative"
            onClick={() => setIsCartOpen(!isCartOpen)}
            aria-label="Shopping cart"
          >
            <ShoppingBag />
            {totalItems > 0 && (
              <span className="absolute -top-1 -right-1 bg-primary text-primary-foreground text-xs rounded-full h-5 w-5 flex items-center justify-center">
                {totalItems}
              </span>
            )}
          </Button>

          {/* Cart Dropdown */}
          {isCartOpen && <CartDropdown onClose={() => setIsCartOpen(false)} />}
        </div>
      </div>

      {/* Mobile Navigation */}
      {isMobileMenuOpen && (
        <div className="md:hidden py-4 border-t">
          <nav className="flex flex-col space-y-4">
            <Link
              to="/"
              className={`hover:text-primary transition-colors ${location.pathname === "/" ? "text-primary font-medium" : ""}`}
            >
              Home
            </Link>
            <Link
              to="/shop"
              className={`hover:text-primary transition-colors ${location.pathname === "/shop" ? "text-primary font-medium" : ""}`}
            >
              Shop
            </Link>
          </nav>
        </div>
      )}
    </header>
  );
}
