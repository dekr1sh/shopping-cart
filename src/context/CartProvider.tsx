import { useState, useEffect, type ReactNode } from "react"
import { CartContext, type CartItem } from "./CartContext.ts"
import type { Product } from "../data/products.ts"

export const CartProvider = ({ children }: { children: ReactNode }) => {
  const [cartItems, setCartItems] = useState<CartItem[]>([])
  const [totalItems, setTotalItems] = useState(0)
  const [totalPrice, setTotalPrice] = useState(0)
  const [isOpen, setIsOpen] = useState(false)

  useEffect(() => {
    const savedCart = localStorage.getItem("cart")
    if (savedCart) {
      setCartItems(JSON.parse(savedCart))
    }
  }, [])

  useEffect(() => {
    localStorage.setItem("cart", JSON.stringify(cartItems))

    const totItems = cartItems.reduce((total, item) => total + item.quantity, 0)
    setTotalItems(totItems)

    const price = cartItems.reduce((total, item) => total + item.price * item.quantity, 0)
    setTotalPrice(price)
  }, [cartItems])

  const addToCart = (product: Product, size: string, quantity: number) => {
    setCartItems((prevItems) => {
      const existingItemIndex = prevItems.findIndex((item) => item.id === product.id && item.size === size)

      if (existingItemIndex >= 0) {
        const updatedItems = [...prevItems]
        updatedItems[existingItemIndex].quantity += quantity
        return updatedItems
      } else {
        return [...prevItems, { ...product, size, quantity }]
      }
    })
  }

  const removeFromCart = (productId: string) => {
    setCartItems((prevItems) => prevItems.filter((item) => item.id !== +productId))
  }

  const updateQuantity = (productId: string, quantity: number) => {
    setCartItems((prevItems) =>
      prevItems.map((item) => (item.id === +productId ? { ...item, quantity: Math.max(1, quantity) } : item)),
    )
  }

  const clearCart = () => {
    setCartItems([])
  }

  return (
    <CartContext.Provider
      value={{ cartItems, totalItems, totalPrice, isOpen, setIsOpen, addToCart, removeFromCart, updateQuantity, clearCart }}
    >
      {children}
    </CartContext.Provider>
  )
}
