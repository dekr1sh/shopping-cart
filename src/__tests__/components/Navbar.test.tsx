import { describe, it, expect, vi, Mock } from "vitest";
import { screen, within } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { render } from "@/__tests__/test-utils";
import Navbar from "@/components/Navbar";
import { useCart } from "@/context/useCart";

vi.mock("@/context/useCart", () => ({ useCart: vi.fn() }));

vi.mock("@/components/CartDropdown", () => ({
  default: vi.fn(() => <div data-testid="cart-dropdown">Mocked CartDropdown</div>),
}));

describe("Navbar", () => {
  it("renders correctly when mobile menu is closed", () => {
    (useCart as Mock).mockReturnValue({ cartItems: [] });
    const { container } = render(<Navbar />);
    expect(container).toMatchSnapshot();
  });

  it("renders correctly when mobile menu is open", async () => {
    (useCart as Mock).mockReturnValue({ cartItems: [] });
    const user = userEvent.setup();
    const { container } = render(<Navbar />);
    const openMenuButton = screen.getByLabelText("Open menu");
  
    await user.click(openMenuButton);      // this is an 'async' action and it is automatically wrapped inside of act() by the RTL
    expect(container).toMatchSnapshot();
  });

  it("toggles mobile menu on button click", async () => {
    (useCart as Mock).mockReturnValue({ cartItems: [] });
    const user = userEvent.setup();

    render(<Navbar />);
    const openMenuButton = screen.getByLabelText("Open menu");
    
    await user.click(openMenuButton);
    expect(screen.queryByTestId("mobile-menu")).toBeInTheDocument();

    const closeMenuButton = screen.getByLabelText("Close menu");
    await user.click(closeMenuButton);
    expect(screen.queryByTestId("mobile-menu")).not.toBeInTheDocument();
  });

  it("closes the mobile menu when navigating to a new route", async () => {
    (useCart as Mock).mockReturnValue({ cartItems: [] });
    const user = userEvent.setup();
    render(<Navbar />);
  
    const menuButton = screen.getByLabelText("Open menu");
    await user.click(menuButton);

    const mobileMenu = screen.getByTestId("mobile-menu");
    const shopLink = within(mobileMenu).getByRole("link", { name: /shop/i });
  
    await user.click(shopLink);
  
    expect(screen.queryByTestId("mobile-menu")).not.toBeInTheDocument();
  });
  
  it("toggles the CartDropdown on cart button click", async () => {
    (useCart as Mock).mockReturnValue({ cartItems: [] });
    const user = userEvent.setup();
    render(<Navbar />);
  
    const cartButton = screen.getByLabelText("Shopping cart");
  
    await user.click(cartButton);
    expect(screen.getByTestId("cart-dropdown")).toBeInTheDocument();
  
    await user.click(cartButton);
    expect(screen.queryByTestId("cart-dropdown")).not.toBeInTheDocument();
  });

  it("closes cart dropdown when clicking outside, but not when clicking inside", async () => {
    (useCart as Mock).mockReturnValue({ cartItems: [] });
    render(<Navbar />);
    const cartButton = screen.getByLabelText("Shopping cart");
  
    await userEvent.click(cartButton);
  
    await userEvent.click(screen.getByTestId("cart-dropdown"));
    expect(screen.getByTestId("cart-dropdown")).toBeInTheDocument();
  
    await userEvent.click(document.body);
    expect(screen.queryByTestId("cart-dropdown")).not.toBeInTheDocument();
  });
  
  it("shows correct item count in cart badge", async () => {
    (useCart as Mock).mockReturnValue({
      cartItems: [
        {
          id: 1,
          name: "Nike Air Max",
          brand: "Nike",
          size: "10",
          quantity: 2,
          price: 150,
          image: "/nike-air-max.jpg",
        },
      ]
    });
    render(<Navbar />);
    const cartBadge = screen.getByText("2"); 
    expect(cartBadge).toBeInTheDocument();
  });
});
