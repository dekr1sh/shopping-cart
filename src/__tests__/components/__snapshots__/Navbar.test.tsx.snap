// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Navbar > renders correctly when mobile menu is closed 1`] = `
<div>
  <header
    class="sticky top-0 z-50 bg-background/80 backdrop-blur-md border-b"
  >
    <div
      class="mx-auto px-4"
    >
      <div
        class="flex items-center justify-between h-16"
      >
        <a
          class="font-bold text-xl flex items-center"
          data-discover="true"
          href="/"
        >
          <span
            class="text-primary"
          >
            Drip
          </span>
          Kicks
        </a>
        <button
          aria-label="Open menu"
          class="md:hidden cursor-pointer"
        >
          <svg
            class="lucide lucide-menu"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <line
              x1="4"
              x2="20"
              y1="12"
              y2="12"
            />
            <line
              x1="4"
              x2="20"
              y1="6"
              y2="6"
            />
            <line
              x1="4"
              x2="20"
              y1="18"
              y2="18"
            />
          </svg>
        </button>
        <nav
          class="hidden md:flex items-center space-x-8"
        >
          <a
            class="hover:text-primary transition-colors text-primary font-medium"
            data-discover="true"
            href="/"
          >
            Home
          </a>
          <a
            class="hover:text-primary transition-colors "
            data-discover="true"
            href="/shop"
          >
            Shop
          </a>
        </nav>
        <div
          class="relative"
        >
          <button
            aria-label="Shopping cart"
            class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive hover:bg-accent hover:text-accent-foreground dark:hover:bg-accent/50 size-9 relative cursor-pointer"
            data-slot="button"
          >
            <svg
              class="lucide lucide-shopping-bag"
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M6 2 3 6v14a2 2 0 0 0 2 2h14a2 2 0 0 0 2-2V6l-3-4Z"
              />
              <path
                d="M3 6h18"
              />
              <path
                d="M16 10a4 4 0 0 1-8 0"
              />
            </svg>
          </button>
        </div>
      </div>
    </div>
  </header>
     
</div>
`;

exports[`Navbar > renders correctly when mobile menu is open 1`] = `
<div>
  <header
    class="sticky top-0 z-50 bg-background/80 backdrop-blur-md border-b"
  >
    <div
      class="mx-auto px-4"
    >
      <div
        class="flex items-center justify-between h-16"
      >
        <a
          class="font-bold text-xl flex items-center"
          data-discover="true"
          href="/"
        >
          <span
            class="text-primary"
          >
            Drip
          </span>
          Kicks
        </a>
        <button
          aria-label="Close menu"
          class="md:hidden cursor-pointer"
        >
          <svg
            class="lucide lucide-x"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M18 6 6 18"
            />
            <path
              d="m6 6 12 12"
            />
          </svg>
        </button>
        <nav
          class="hidden md:flex items-center space-x-8"
        >
          <a
            class="hover:text-primary transition-colors text-primary font-medium"
            data-discover="true"
            href="/"
          >
            Home
          </a>
          <a
            class="hover:text-primary transition-colors "
            data-discover="true"
            href="/shop"
          >
            Shop
          </a>
        </nav>
        <div
          class="relative"
        >
          <button
            aria-label="Shopping cart"
            class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive hover:bg-accent hover:text-accent-foreground dark:hover:bg-accent/50 size-9 relative cursor-pointer"
            data-slot="button"
          >
            <svg
              class="lucide lucide-shopping-bag"
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M6 2 3 6v14a2 2 0 0 0 2 2h14a2 2 0 0 0 2-2V6l-3-4Z"
              />
              <path
                d="M3 6h18"
              />
              <path
                d="M16 10a4 4 0 0 1-8 0"
              />
            </svg>
          </button>
        </div>
      </div>
      <div
        class="md:hidden py-4 border-t"
        data-testid="mobile-menu"
      >
        <nav
          class="flex flex-col space-y-4"
        >
          <a
            class="hover:text-primary transition-colors text-primary font-medium"
            data-discover="true"
            href="/"
          >
            Home
          </a>
          <a
            class="hover:text-primary transition-colors "
            data-discover="true"
            href="/shop"
          >
            Shop
          </a>
        </nav>
      </div>
    </div>
  </header>
     
</div>
`;
